File: generated-code/codecraft/scala/src/main/scala/model/EntityProperties.scala
package model

import util.StreamUtil

/**
 * Entity properties
 *
 * @param size Size. Entity has a form of a square with side of this length
 * @param buildScore Score for building this entity
 * @param destroyScore Score for destroying this entity
 * @param canMove Whether this entity can move
 * @param populationProvide Number of population points this entity provides, if active
 * @param populationUse Number of population points this entity uses
 * @param maxHealth Maximum health points
 * @param initialCost Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type
 * @param sightRange If fog of war is enabled, maximum distance at which other entities are considered visible
 * @param resourcePerHealth Amount of resource added to enemy able to collect resource on dealing damage for 1 health point
 * @param build Build properties, if entity can build
 * @param attack Attack properties, if entity can attack
 * @param repair Repair properties, if entity can repair
 */
case class EntityProperties(size: Int, buildScore: Int, destroyScore: Int, canMove: Boolean, populationProvide: Int, populationUse: Int, maxHealth: Int, initialCost: Int, sightRange: Int, resourcePerHealth: Int, build: Option[model.BuildProperties], attack: Option[model.AttackProperties], repair: Option[model.RepairProperties]) {
    /**
     * Write EntityProperties to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, size)
        StreamUtil.writeInt(stream, buildScore)
        StreamUtil.writeInt(stream, destroyScore)
        StreamUtil.writeBoolean(stream, canMove)
        StreamUtil.writeInt(stream, populationProvide)
        StreamUtil.writeInt(stream, populationUse)
        StreamUtil.writeInt(stream, maxHealth)
        StreamUtil.writeInt(stream, initialCost)
        StreamUtil.writeInt(stream, sightRange)
        StreamUtil.writeInt(stream, resourcePerHealth)
        build match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
        attack match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
        repair match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
    }

    /**
     * Get string representation of EntityProperties
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("EntityProperties { ")
        stringBuilder.append("size: ")
        stringBuilder.append(size)
        stringBuilder.append(", ")
        stringBuilder.append("buildScore: ")
        stringBuilder.append(buildScore)
        stringBuilder.append(", ")
        stringBuilder.append("destroyScore: ")
        stringBuilder.append(destroyScore)
        stringBuilder.append(", ")
        stringBuilder.append("canMove: ")
        stringBuilder.append(canMove)
        stringBuilder.append(", ")
        stringBuilder.append("populationProvide: ")
        stringBuilder.append(populationProvide)
        stringBuilder.append(", ")
        stringBuilder.append("populationUse: ")
        stringBuilder.append(populationUse)
        stringBuilder.append(", ")
        stringBuilder.append("maxHealth: ")
        stringBuilder.append(maxHealth)
        stringBuilder.append(", ")
        stringBuilder.append("initialCost: ")
        stringBuilder.append(initialCost)
        stringBuilder.append(", ")
        stringBuilder.append("sightRange: ")
        stringBuilder.append(sightRange)
        stringBuilder.append(", ")
        stringBuilder.append("resourcePerHealth: ")
        stringBuilder.append(resourcePerHealth)
        stringBuilder.append(", ")
        stringBuilder.append("build: ")
        stringBuilder.append(build)
        stringBuilder.append(", ")
        stringBuilder.append("attack: ")
        stringBuilder.append(attack)
        stringBuilder.append(", ")
        stringBuilder.append("repair: ")
        stringBuilder.append(repair)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object EntityProperties {
    /**
     * Read EntityProperties from input stream
     */
    def readFrom(stream: java.io.InputStream): EntityProperties = EntityProperties(
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readBoolean(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        if (StreamUtil.readBoolean(stream)) Some(
            model.BuildProperties.readFrom(stream)
        ) else None,
        if (StreamUtil.readBoolean(stream)) Some(
            model.AttackProperties.readFrom(stream)
        ) else None,
        if (StreamUtil.readBoolean(stream)) Some(
            model.RepairProperties.readFrom(stream)
        ) else None
    )
}

File: generated-code/codecraft/cpp/model/EntityProperties.cpp
#include "EntityProperties.hpp"

EntityProperties::EntityProperties() { }

EntityProperties::EntityProperties(int size, int buildScore, int destroyScore, bool canMove, int populationProvide, int populationUse, int maxHealth, int initialCost, int sightRange, int resourcePerHealth, std::optional<BuildProperties> build, std::optional<AttackProperties> attack, std::optional<RepairProperties> repair) : size(size), buildScore(buildScore), destroyScore(destroyScore), canMove(canMove), populationProvide(populationProvide), populationUse(populationUse), maxHealth(maxHealth), initialCost(initialCost), sightRange(sightRange), resourcePerHealth(resourcePerHealth), build(build), attack(attack), repair(repair) { }

// Read EntityProperties from input stream
EntityProperties EntityProperties::readFrom(InputStream& stream) {
    int size;
    size = stream.readInt();
    int buildScore;
    buildScore = stream.readInt();
    int destroyScore;
    destroyScore = stream.readInt();
    bool canMove;
    canMove = stream.readBool();
    int populationProvide;
    populationProvide = stream.readInt();
    int populationUse;
    populationUse = stream.readInt();
    int maxHealth;
    maxHealth = stream.readInt();
    int initialCost;
    initialCost = stream.readInt();
    int sightRange;
    sightRange = stream.readInt();
    int resourcePerHealth;
    resourcePerHealth = stream.readInt();
    std::optional<BuildProperties> build;
    if (stream.readBool()) {
        BuildProperties buildValue;
        buildValue = BuildProperties::readFrom(stream);
        build = buildValue;
    } else {
        build = std::optional<BuildProperties>();
    }
    std::optional<AttackProperties> attack;
    if (stream.readBool()) {
        AttackProperties attackValue;
        attackValue = AttackProperties::readFrom(stream);
        attack = attackValue;
    } else {
        attack = std::optional<AttackProperties>();
    }
    std::optional<RepairProperties> repair;
    if (stream.readBool()) {
        RepairProperties repairValue;
        repairValue = RepairProperties::readFrom(stream);
        repair = repairValue;
    } else {
        repair = std::optional<RepairProperties>();
    }
    return EntityProperties(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, initialCost, sightRange, resourcePerHealth, build, attack, repair);
}

// Write EntityProperties to output stream
void EntityProperties::writeTo(OutputStream& stream) const {
    stream.write(size);
    stream.write(buildScore);
    stream.write(destroyScore);
    stream.write(canMove);
    stream.write(populationProvide);
    stream.write(populationUse);
    stream.write(maxHealth);
    stream.write(initialCost);
    stream.write(sightRange);
    stream.write(resourcePerHealth);
    if (build) {
        stream.write(true);
        const BuildProperties& buildValue = *build;
        buildValue.writeTo(stream);
    } else {
        stream.write(false);
    }
    if (attack) {
        stream.write(true);
        const AttackProperties& attackValue = *attack;
        attackValue.writeTo(stream);
    } else {
        stream.write(false);
    }
    if (repair) {
        stream.write(true);
        const RepairProperties& repairValue = *repair;
        repairValue.writeTo(stream);
    } else {
        stream.write(false);
    }
}

// Get string representation of EntityProperties
std::string EntityProperties::toString() const {
    std::stringstream ss;
    ss << "EntityProperties { ";
    ss << "size: ";
    ss << size;
    ss << ", ";
    ss << "buildScore: ";
    ss << buildScore;
    ss << ", ";
    ss << "destroyScore: ";
    ss << destroyScore;
    ss << ", ";
    ss << "canMove: ";
    ss << canMove;
    ss << ", ";
    ss << "populationProvide: ";
    ss << populationProvide;
    ss << ", ";
    ss << "populationUse: ";
    ss << populationUse;
    ss << ", ";
    ss << "maxHealth: ";
    ss << maxHealth;
    ss << ", ";
    ss << "initialCost: ";
    ss << initialCost;
    ss << ", ";
    ss << "sightRange: ";
    ss << sightRange;
    ss << ", ";
    ss << "resourcePerHealth: ";
    ss << resourcePerHealth;
    ss << ", ";
    ss << "build: ";
    if (build) {
        const BuildProperties& buildValue = *build;
        ss << buildValue.toString();
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "attack: ";
    if (attack) {
        const AttackProperties& attackValue = *attack;
        ss << attackValue.toString();
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "repair: ";
    if (repair) {
        const RepairProperties& repairValue = *repair;
        ss << repairValue.toString();
    } else {
        ss << "none";
    }
    ss << " }";
    return ss.str();
}